solutions/gold/cses-1665.mdx

```diff
@@ -2,413 +2,96 @@
-	DBG(u...);
-}
-#ifdef LOCAL  // compile with -DLOCAL, chk -> fake assert
-#define dbg(...)                                                               \
-	cerr << "Line(" << __LINE__ << ") -> [" << #__VA_ARGS__ << "]: [",         \
-	    DBG(__VA_ARGS__)
-#define chk(...)                                                               \
-	if (!(__VA_ARGS__))                                                        \
-		cerr << "Line(" << __LINE__ << ") -> function(" << __FUNCTION__        \
-		     << ") -> CHK FAILED: (" << #__VA_ARGS__ << ")"                    \
-		     << "\n",                                                          \
-		    exit(0);
-#else
-#define dbg(...) 0
-#define chk(...) 0
-#endif
-
-void setPrec() { cout << fixed << setprecision(15); }
-void unsyncIO() { cin.tie(0)->sync_with_stdio(0); }
-// FILE I/O
-void setIn(str s) { freopen(s.c_str(), "r", stdin); }
-void setOut(str s) { freopen(s.c_str(), "w", stdout); }
-void setIO(str s = "") {
-	unsyncIO();
-	setPrec();
-	// cin.exceptions(cin.failbit);
-	// throws exception when do smth illegal
-	// ex. try to read letter into int
-	if (sz(s)) setIn(s + ".in"), setOut(s + ".out");  // for USACO
-}
-
-/**
- * Description: modular arithmetic operations
- * Source:
- * KACTL
- * https://codeforces.com/blog/entry/63903
- * https://codeforces.com/contest/1261/submission/65632855 (tourist)
- * https://codeforces.com/contest/1264/submission/66344993 (ksun)
- * also see https://github.com/ecnerwala/cp-book/blob/master/src/modnum.hpp
- * (ecnerwal) Verification: https://open.kattis.com/problems/modulararithmetic
- */
 
-template <int MOD, int RT> struct mint {
-	static const int mod = MOD;
-	static constexpr mint rt() { return RT; }  // primitive root for FFT
-	int v;
-	explicit operator int() const {
-		return v;
-	}  // explicit -> don't silently convert to int
-	mint() { v = 0; }
-	mint(ll _v) {
-		v = int((-MOD < _v && _v < MOD) ? _v : _v % MOD);
-		if (v < 0) v += MOD;
-	}
-	friend bool operator==(const mint &a, const mint &b) { return a.v == b.v; }
-	friend bool operator!=(const mint &a, const mint &b) { return !(a == b); }
-	friend bool operator<(const mint &a, const mint &b) { return a.v < b.v; }
-	friend void re(mint &a) {
-		ll x;
-		re(x);
-		a = mint(x);
-	}
-	friend str ts(mint a) { return ts(a.v); }
+const int N = 101;
+const int X = 5e3 + 1;
+const int M = 1e9 + 7;
+const int K = 5e3;  // K is an offset to account for negatives
+int n, x, s[N];     // x -> max penalty
+
+// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
+// for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
+// lines up with s[i] (i.e. skill of person i + 1 = s[i])
+ll dp1[N][X + K];  // dp[i][j][k]
+ll dp2[N][X + K];  // dp[i + 1][j][k]
+// Naively, our dp table would require N * N * X memory
+// However, dp[i][j][k] can only affect dp[i + 1][x][y],
+// so we effectively only need to store two rows of the dp table.
```

Comment:
```suggestion
// Notice we only need to store 2 rows!
ll dp1[N][X + K];  // The subarray we're currently at
ll dp2[N][X + K];  // The next subarray (we'll fill this up with dp1)
```

Reply:
i hope my understanding of your code is correct

Reply:
yeah it is :)